framework/base/Security.php

```diff
@@ -472,71 +479,61 @@ public function generateRandomKey($length = 32)
-        if ($this->_randomSource === self::SOURCE_MCRYPT
-            || ($this->_randomSource === null
-                && PHP_VERSION_ID >= 50307
-                && function_exists('mcrypt_create_iv'))
-        ) {
+        // CryptGenRandom on Windows. Elsewhere it directly reads /dev/urandom.
+        if (function_exists('mcrypt_create_iv')) {
             $key = mcrypt_create_iv($length, MCRYPT_DEV_URANDOM);
             if (StringHelper::byteLength($key) === $length) {
-                $this->_randomSource = self::SOURCE_MCRYPT;
-
                 return $key;
             }
-
-            $this->_randomSource = null;
         }
 
-        // If not on Windows, test for a /dev/urandom device.
-        if ($this->_randomSource === null && DIRECTORY_SEPARATOR === '/') {
-            // Check it for speacial character device protection mode. Do not follow
-            // symbolic link at '/dev/urandom', as such would be suspicious. With lstat()
-            // (as opposed to stat()) the test fails if it is.
-            $lstat = @lstat(self::DEV_URANDOM);
-            $urandomDevice = $lstat !== false && ($lstat['mode'] & 0170000) === 020000;
-        } else {
-            $urandomDevice = false;
-        }
-        if ($this->_randomSource === self::SOURCE_URANDOM || $urandomDevice) {
-            $key = @file_get_contents(self::DEV_URANDOM, false, null, 0, $length);
-
-            if ($key !== false && StringHelper::byteLength($key) === $length) {
-                $this->_randomSource = self::SOURCE_URANDOM;
-
-                return $key;
+        // If not on Windows, try to open a random device.
+        if ($this->_randomFile === null && DIRECTORY_SEPARATOR === '/') {
+            // urandom is a symlink to random on FreeBSD.
+            $device = PHP_OS === 'FreeBSD' ? '/dev/random' : '/dev/urandom';
+            // Check random device for special character device protection mode. Use lstat()
+            // instead of stat() in case an attacker arranges a symlink to a fake device.
+            $lstat = @lstat($device);
+            if ($lstat !== false && ($lstat['mode'] & 0170000) === 020000) {
+                $this->_randomFile = fopen($device, 'rb') ?: null;
+
+                if (is_resource($this->_randomFile)) {
+                    // By default PHP buffer size is 8192 bytes which causes wasting
+                    // more entropy that we're actually using. Therefore setting it to
+                    // lower value.
+                    if (function_exists('stream_set_read_buffer')) {
+                        stream_set_read_buffer($this->_randomFile, self::RANDOM_FILE_BUFFER);
+                    }
+                    // stream_set_read_buffer() isn't implemented on HHVM
+                    if (function_exists('stream_set_chunk_size')) {
+                        stream_set_chunk_size($this->_randomFile, self::RANDOM_FILE_BUFFER);
+                    }
```

Comment:
Yes. That's for file streams at least. The function sets buffer size for a file pointer specified.